File: scripts/report.py
"""
report.py - Research pack generator for SNI Research Tool

Reads verified articles, groups by sector, checks off-limits,
and generates a markdown research pack.

Usage:
  python scripts/report.py --test
  python scripts/report.py --start-date 2026-02-13 --end-date 2026-02-20
"""

import os
import sys
import io
import json
import argparse
from datetime import datetime, timedelta
from pathlib import Path

# Fix Windows console encoding for Unicode output
if sys.stdout.encoding and sys.stdout.encoding.lower() not in ('utf-8', 'utf8'):
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')
if sys.stderr.encoding and sys.stderr.encoding.lower() not in ('utf-8', 'utf8'):
    sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8', errors='replace')

import yaml

ROOT = Path(__file__).parent.parent
sys.path.insert(0, str(ROOT))

with open(ROOT / 'config' / 'sectors.yaml', 'r', encoding='utf-8') as f:
    SECTORS_CONFIG = yaml.safe_load(f)['sectors']

with open(ROOT / 'config' / 'off-limits.yaml', 'r', encoding='utf-8') as f:
    OFF_LIMITS = yaml.safe_load(f)

SECTOR_ORDER = ['general', 'biopharma', 'medtech', 'manufacturing', 'insurance']
SECTOR_DISPLAY = {
    'general': 'General AI',
    'biopharma': 'Pharma & Biopharma',
    'medtech': 'MedTech',
    'manufacturing': 'Complex & Advanced Manufacturing',
    'insurance': 'Insurance',
}


def get_date_window(args):
    today = datetime.now().date()
    if args.test:
        start = today - timedelta(days=7)
        return str(start), str(today)
    if args.start_date and args.end_date:
        return args.start_date, args.end_date
    d = str(today)
    return d, d


def get_week_number(date_str):
    d = datetime.strptime(date_str, '%Y-%m-%d')
    return d.isocalendar()[1]


def format_date(date_str):
    try:
        d = datetime.strptime(date_str, '%Y-%m-%d')
        return d.strftime('%-d %B %Y') if os.name != 'nt' else d.strftime('%d %B %Y').lstrip('0')
    except Exception:
        return date_str


def load_verified_articles(start_date, end_date):
    """Load all verified article JSONs within the date window."""
    verified_dir = ROOT / 'data' / 'verified'
    if not verified_dir.exists():
        return []

    articles = []
    for date_dir in verified_dir.iterdir():
        if not date_dir.is_dir():
            continue
        if not (start_date <= date_dir.name <= end_date):
            continue
        for sector_dir in date_dir.iterdir():
            if not sector_dir.is_dir():
                continue
            for json_file in sector_dir.glob('*.json'):
                try:
                    article = json.loads(json_file.read_text(encoding='utf-8'))
                    article.pop('_raw_html', None)
                    articles.append(article)
                except Exception:
                    continue

    # Sort: date descending, then title ascending
    articles.sort(key=lambda a: (a.get('date_published', ''), a.get('title', '')),
                  reverse=False)
    articles.sort(key=lambda a: a.get('date_published', ''), reverse=True)
    return articles


def deduplicate(articles):
    seen = set()
    result = []
    for a in articles:
        url = a.get('url', '')
        if url not in seen:
            seen.add(url)
            result.append(a)
    return result


def group_by_sector(articles):
    groups = {}
    for a in articles:
        sector = a.get('sector', 'general')
        groups.setdefault(sector, []).append(a)
    return groups


def check_off_limits_report(articles):
    """Return list of conflicts for reporting."""
    conflicts = []
    for article in articles:
        search_text = f'{article.get("title", "")} {article.get("snippet", "")}'.lower()
        for week, entries in OFF_LIMITS.items():
            if not isinstance(entries, list):
                continue
            for entry in entries:
                company = (entry.get('company') or '').lower()
                topic = (entry.get('topic') or '').lower()
                if not company or company not in search_text:
                    continue
                topic_words = [w for w in topic.split() if len(w) > 3]
                if topic_words:
                    matches = sum(1 for w in topic_words if w in search_text)
                    if matches >= max(1, len(topic_words) // 2):
                        conflicts.append({
                            'article': article.get('title', ''),
                            'reason': f'{week}: {entry.get("company")} - {entry.get("topic")}',
                        })
    return conflicts


def format_sector_section(sector_articles, display_name):
    """Format a sector's articles as markdown."""
    if not sector_articles:
        return ''
    lines = [f'## {display_name} ({len(sector_articles)} article{"s" if len(sector_articles) != 1 else ""}, all dates verified)\n']
    for i, a in enumerate(sector_articles, 1):
        title = a.get('title', 'Untitled')
        url = a.get('url', '')
        source = a.get('source', '')
        date = a.get('date_published', '')
        method = a.get('date_verified_method', '')
        confidence = a.get('date_confidence', '')
        snippet = (a.get('snippet') or '').replace('\n', ' ').strip()[:300]

        lines.append(f'### Story {i}: {title}')
        lines.append(f'- Source: {source}')
        lines.append(f'- URL: {url}')
        lines.append(f'- Published: {format_date(date)} (verified: {method}, confidence: {confidence})')
        lines.append(f'- Summary: {snippet}')
        lines.append('')
    return '\n'.join(lines)


def generate_report(articles, start_date, end_date):
    """Generate the full research pack markdown."""
    deduped = deduplicate(articles)
    grouped = group_by_sector(deduped)
    week_num = get_week_number(end_date)
    date_range = f'{format_date(start_date)} - {format_date(end_date)}'
    generated = datetime.now().strftime('%d %B %Y %H:%M')
    weeks_checked = ', '.join(k for k in OFF_LIMITS if isinstance(OFF_LIMITS[k], list))

    off_limit_conflicts = check_off_limits_report(deduped)

    lines = [
        f'# SNI Research Pack: Week {week_num}, 2026',
        f'Generated: {generated}',
        f'Date range: {date_range}',
        f'Total verified articles: {len(deduped)}',
        '',
        '---',
        '',
    ]

    # Headlines overview
    lines.append('## Headlines Overview')
    lines.append('')
    for sector in SECTOR_ORDER:
        sector_articles = grouped.get(sector, [])
        if not sector_articles:
            continue
        display = SECTOR_DISPLAY.get(sector, sector)
        lines.append(f'**{display}** ({len(sector_articles)})')
        for a in sector_articles:
            lines.append(f'• {a.get("title", "")} [{a.get("source", "")}, {a.get("date_published", "")}]')
        lines.append('')

    lines.append('---')
    lines.append('')

    # Detailed sector sections
    for sector in SECTOR_ORDER:
        sector_articles = grouped.get(sector, [])
        display = SECTOR_DISPLAY.get(sector, sector)
        section = format_sector_section(sector_articles, display)
        if section:
            lines.append(section)
            lines.append('---')
            lines.append('')

    # Off-limits check
    lines.append('## Off-Limits Check')
    lines.append(f'- Checked against: {weeks_checked}')
    if not off_limit_conflicts:
        lines.append('- Conflicts found: 0 ✓')
    else:
        lines.append(f'- Conflicts found: {len(off_limit_conflicts)} ⚠')
        for c in off_limit_conflicts:
            lines.append(f'  - "{c["article"]}" → {c["reason"]}')
    lines.append('')

    # Stats
    lines.append('## Collection Statistics')
    for sector in SECTOR_ORDER:
        count = len(grouped.get(sector, []))
        display = SECTOR_DISPLAY.get(sector, sector)
        lines.append(f'- {display}: {count} articles')

    return '\n'.join(lines), {
        'total': len(deduped),
        'by_sector': {s: len(grouped.get(s, [])) for s in SECTOR_ORDER},
    }


def main():
    parser = argparse.ArgumentParser(description='SNI Research Report Generator')
    parser.add_argument('--test', action='store_true')
    parser.add_argument('--start-date')
    parser.add_argument('--end-date')
    args = parser.parse_args()

    start_date, end_date = get_date_window(args)

    print()
    print('═' * 49)
    print('  SNI Research Tool - Report')
    print(f'  Date window: {start_date} → {end_date}')
    print('═' * 49)
    print()

    articles = load_verified_articles(start_date, end_date)
    print(f'Found {len(articles)} verified articles in window')

    if not articles:
        print('No articles found. Run fetch.py first.')
        return

    report_md, stats = generate_report(articles, start_date, end_date)

    # Save report
    output_dir = ROOT / 'output'
    output_dir.mkdir(exist_ok=True)
    week_num = get_week_number(end_date)
    suffix = 'test' if args.test else f'week-{week_num}'
    report_path = output_dir / f'{end_date}-{suffix}-research.md'
    report_path.write_text(report_md, encoding='utf-8')

    print()
    print('═' * 49)
    print(f'  Report saved: {report_path}')
    print()
    print('  Articles by sector:')
    for sector, count in stats['by_sector'].items():
        if count > 0:
            display = SECTOR_DISPLAY.get(sector, sector)
            print(f'    {display:<30} {count}')
    print(f'    {"TOTAL":<30} {stats["total"]}')
    print('═' * 49)
    print()

    # Print headline summary to console
    print('── Headline Summary ' + '─' * 30)
    deduped = deduplicate(articles)
    grouped = group_by_sector(deduped)
    for sector in SECTOR_ORDER:
        sector_articles = grouped.get(sector, [])
        if not sector_articles:
            continue
        display = SECTOR_DISPLAY.get(sector, sector)
        print(f'\n{display} ({len(sector_articles)}):')
        for a in sector_articles:
            print(f'  • {a.get("title", "")[:80]} [{a.get("date_published", "")}]')
    print()


if __name__ == '__main__':
    main()

File: scripts/report.js
/**
 * report.js - Research pack generator for SNI Research Tool
 *
 * Reads verified articles, groups by sector, checks off-limits,
 * and generates a markdown research pack.
 *
 * Usage:
 *   bun scripts/report.js --test              # Last 7 days
 *   bun scripts/report.js --week 9            # ISO week 9 of current year
 *   bun scripts/report.js --week 9 --year 2026
 *   bun scripts/report.js --start-date 2026-02-13 --end-date 2026-02-20
 */

import { readFileSync, writeFileSync, readdirSync, mkdirSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import yaml from 'js-yaml';
import { getISOWeekNumber, getISOYearForWeek, getWeekWindow } from './lib/week.js';
import { loadSectorNames } from './lib/prompt.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, '..');

const sectorsConfig = yaml.load(readFileSync(join(ROOT, 'config', 'sectors.yaml'), 'utf8'));
const offLimits = yaml.load(readFileSync(join(ROOT, 'config', 'off-limits.yaml'), 'utf8'));

// Canonical sector ordering and display names from config
const sectorNamesConfig = loadSectorNames();
const SECTOR_ORDER = Object.entries(sectorNamesConfig)
  .sort((a, b) => a[1].order - b[1].order)
  .map(([key]) => key);
const SECTOR_DISPLAY_NAMES = Object.fromEntries(
  Object.entries(sectorNamesConfig).map(([key, val]) => [key, val.config])
);

function parseArgs(argv) {
  const args = {};
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--test') args.test = true;
    if (argv[i] === '--start-date') args.startDate = argv[++i];
    if (argv[i] === '--end-date') args.endDate = argv[++i];
    if (argv[i] === '--week') args.week = parseInt(argv[++i], 10);
    if (argv[i] === '--year') args.year = parseInt(argv[++i], 10);
  }
  return args;
}

function getDateWindow(args) {
  if (args.week) {
    const year = args.year || new Date().getFullYear();
    const { start, end } = getWeekWindow(args.week, year);
    return { startDate: start, endDate: end };
  }
  const today = new Date();
  if (args.test) {
    const start = new Date(today);
    start.setDate(start.getDate() - 7);
    return {
      startDate: start.toISOString().slice(0, 10),
      endDate: today.toISOString().slice(0, 10),
    };
  }
  if (args.startDate && args.endDate) {
    return { startDate: args.startDate, endDate: args.endDate };
  }
  const d = today.toISOString().slice(0, 10);
  return { startDate: d, endDate: d };
}

function getAllVerifiedArticles(window) {
  const verifiedDir = join(ROOT, 'data', 'verified');
  if (!existsSync(verifiedDir)) return [];

  const articles = [];
  const dateDirs = readdirSync(verifiedDir, { withFileTypes: true })
    .filter(d => d.isDirectory() && d.name >= window.startDate && d.name <= window.endDate)
    .map(d => d.name);

  for (const dateDir of dateDirs) {
    const datePath = join(verifiedDir, dateDir);
    const sectorDirs = readdirSync(datePath, { withFileTypes: true })
      .filter(d => d.isDirectory())
      .map(d => d.name);

    for (const sectorDir of sectorDirs) {
      const sectorPath = join(datePath, sectorDir);
      const files = readdirSync(sectorPath).filter(f => f.endsWith('.json'));
      for (const file of files) {
        try {
          const article = JSON.parse(readFileSync(join(sectorPath, file), 'utf8'));
          // Remove raw HTML from report processing
          delete article._raw_html;
          articles.push(article);
        } catch { /* skip corrupt files */ }
      }
    }
  }

  // Sort by date descending, then by title
  return articles.sort((a, b) => {
    if (b.date_published !== a.date_published) return b.date_published.localeCompare(a.date_published);
    return a.title.localeCompare(b.title);
  });
}

function deduplicateArticles(articles) {
  const seen = new Set();
  return articles.filter(a => {
    if (seen.has(a.url)) return false;
    seen.add(a.url);
    return true;
  });
}

function groupBySector(articles) {
  const groups = {};
  for (const article of articles) {
    const sector = article.sector || 'general';
    if (!groups[sector]) groups[sector] = [];
    groups[sector].push(article);
  }
  return groups;
}

function formatDate(dateStr) {
  try {
    const d = new Date(dateStr + 'T12:00:00Z');
    return d.toLocaleDateString('en-GB', { day: 'numeric', month: 'long', year: 'numeric', timeZone: 'UTC' });
  } catch { return dateStr; }
}

// Replaced naive day-of-year/7 math with proper ISO week calculation
function getWeekNumber(dateStr) {
  return getISOWeekNumber(dateStr);
}

function summariseSector(articles, sectorName, displayName) {
  if (!articles || articles.length === 0) return '';

  const lines = [`## ${displayName} (${articles.length} article${articles.length !== 1 ? 's' : ''}, all dates verified)\n`];

  articles.forEach((a, i) => {
    const verifiedNote = `verified: ${a.date_verified_method}`;
    const summary = a.snippet
      ? a.snippet.replace(/\s+/g, ' ').trim().slice(0, 300)
      : '(no snippet available)';

    lines.push(`### Story ${i + 1}: ${a.title}`);
    lines.push(`- Source: ${a.source}`);
    lines.push(`- URL: ${a.url}`);
    lines.push(`- Published: ${formatDate(a.date_published)} (${verifiedNote})`);
    lines.push(`- Confidence: ${a.date_confidence}`);
    lines.push(`- Summary: ${summary}`);
    lines.push('');
  });

  return lines.join('\n');
}

function checkOffLimitsReport(articles, offLimits) {
  const conflicts = [];
  for (const article of articles) {
    const searchText = `${article.title} ${article.snippet || ''}`.toLowerCase();
    for (const [week, entries] of Object.entries(offLimits)) {
      if (!Array.isArray(entries)) continue;
      for (const entry of entries) {
        const company = entry.company?.toLowerCase();
        const topic = entry.topic?.toLowerCase();
        if (!company) continue;
        if (searchText.includes(company)) {
          const topicWords = (topic || '').split(' ').filter(w => w.length > 3);
          const matches = topicWords.filter(w => searchText.includes(w)).length;
          if (matches >= Math.max(1, Math.floor(topicWords.length * 0.5))) {
            conflicts.push({
              article: article.title,
              reason: `${week}: ${entry.company} - ${entry.topic}`,
            });
          }
        }
      }
    }
  }
  return conflicts;
}

function generateReport(articles, window) {
  const deduplicated = deduplicateArticles(articles);
  const grouped = groupBySector(deduplicated);
  const weeksRange = `${formatDate(window.startDate)} - ${formatDate(window.endDate)}`;
  const weekNum = getWeekNumber(window.endDate);
  const weekYear = getISOYearForWeek(window.endDate);

  const offLimitConflicts = checkOffLimitsReport(deduplicated, offLimits);
  const weeksChecked = Object.keys(offLimits).join(', ');

  let report = `# SNI Research Pack: Week ${weekNum}, ${weekYear}
Generated: ${new Date().toLocaleDateString('en-GB', { day: 'numeric', month: 'long', year: 'numeric' })}
Date range: ${weeksRange}
Total verified articles: ${deduplicated.length}

---

`;

  // Headline summary table
  report += `## Headlines Overview\n\n`;
  for (const sector of SECTOR_ORDER) {
    const articles = grouped[sector];
    if (!articles || articles.length === 0) continue;
    const name = SECTOR_DISPLAY_NAMES[sector] || sector;
    report += `**${name}** (${articles.length})\n`;
    for (const a of articles) {
      report += `• ${a.title} [${a.source}, ${a.date_published}]\n`;
    }
    report += '\n';
  }

  report += '---\n\n';

  // Detailed sector sections
  for (const sector of SECTOR_ORDER) {
    const sectorArticles = grouped[sector];
    const displayName = SECTOR_DISPLAY_NAMES[sector] || sector;
    if (sectorArticles && sectorArticles.length > 0) {
      report += summariseSector(sectorArticles, sector, displayName);
      report += '---\n\n';
    }
  }

  // Off-limits check
  report += `## Off-Limits Check\n`;
  report += `- Checked against: ${weeksChecked}\n`;
  if (offLimitConflicts.length === 0) {
    report += `- Conflicts found: 0 ✓\n`;
  } else {
    report += `- Conflicts found: ${offLimitConflicts.length} ⚠\n`;
    for (const conflict of offLimitConflicts) {
      report += `  - "${conflict.article}" → ${conflict.reason}\n`;
    }
  }
  report += '\n';

  // Stats
  report += `## Collection Statistics\n`;
  for (const sector of SECTOR_ORDER) {
    const count = grouped[sector]?.length || 0;
    const name = SECTOR_DISPLAY_NAMES[sector] || sector;
    report += `- ${name}: ${count} articles\n`;
  }

  return { report, stats: { total: deduplicated.length, bySector: Object.fromEntries(SECTOR_ORDER.map(s => [s, grouped[s]?.length || 0])) } };
}

export async function runReport(args = {}) {
  const window = getDateWindow(args);

  console.log('');
  console.log('═══════════════════════════════════════════════');
  console.log('  SNI Research Tool - Report');
  if (args.week) console.log(`  Week mode: Week ${args.week}, ${args.year || new Date().getFullYear()}`);
  console.log(`  Date window: ${window.startDate} → ${window.endDate}`);
  console.log('═══════════════════════════════════════════════');
  console.log('');

  const articles = getAllVerifiedArticles(window);
  console.log(`Found ${articles.length} verified articles in window`);

  if (articles.length === 0) {
    console.log('No articles found. Run fetch.js first.');
    return { total: 0, bySector: {}, reportPath: null };
  }

  const { report, stats } = generateReport(articles, window);

  // Save report
  mkdirSync(join(ROOT, 'output'), { recursive: true });
  const weekNum = getWeekNumber(window.endDate);
  const suffix = args.test ? 'test' : `week-${weekNum}`;
  const reportPath = join(ROOT, 'output', `${window.endDate}-${suffix}-research.md`);
  try {
    writeFileSync(reportPath, report);
  } catch (err) {
    console.warn(`Failed to save report: ${err.message}`);
    return { ...stats, reportPath: null };
  }

  console.log('');
  console.log('═══════════════════════════════════════════════');
  console.log(`  Report saved: ${reportPath}`);
  console.log('');
  console.log('  Articles by sector:');
  for (const [sector, count] of Object.entries(stats.bySector)) {
    if (count > 0) console.log(`    ${sector.padEnd(15)} ${count}`);
  }
  console.log(`    ${'TOTAL'.padEnd(15)} ${stats.total}`);
  console.log('═══════════════════════════════════════════════');
  console.log('');

  // Print headline summary to console
  console.log('── Headline Summary ────────────────────────────');
  const grouped = {};
  for (const a of articles) {
    if (!grouped[a.sector]) grouped[a.sector] = [];
    grouped[a.sector].push(a);
  }
  for (const sector of SECTOR_ORDER) {
    const sArticles = grouped[sector];
    if (!sArticles || sArticles.length === 0) continue;
    console.log(`\n${SECTOR_DISPLAY_NAMES[sector] || sector} (${sArticles.length}):`);
    for (const a of sArticles) {
      console.log(`  • ${a.title.slice(0, 80)} [${a.date_published}]`);
    }
  }
  console.log('');

  return { ...stats, reportPath };
}

if (import.meta.main) {
  const args = parseArgs(process.argv.slice(2));
  runReport(args)
    .then(stats => {
      console.log(`Result: ${stats.total} articles, report: ${stats.reportPath || 'none'}`);
      process.exit(0);
    })
    .catch(e => {
      console.error('Fatal error:', e);
      process.exit(1);
    });
}
